include/deal.II/sundials/arkode.h

```diff
@@ -0,0 +1,858 @@
+     * this function consecutively several times by simply performing all
+     * intermediate interpolations. There is no relationship between how many
+     * times this function is called and how many time steps have actually been
+     * computed.
+     */
+    std::function<void(const double t,
+                       const VectorType &sol,
+                       const unsigned int step_number)> output_step;
+
+    /**
+     * Evaluate wether the solver should be restarted (for example because the
+     * number of degrees of freedom has changed).
+     *
+     * This function is supposed to perform all operations that are necessary
+     * in `sol` to make sure that the resulting vectors are consistent, and of
+     * the correct final size.
+     *
+     * For example, one may decide that a local refinement is necessary at time
+     * t. This function should then return true, and change the dimension of
+     * `sol` to reflect the new dimension. Since ARKode does not know about the
+     * new dimension, an internal reset is necessary.
+     *
+     * The default implementation simply returns `false`, i.e., no restart is
+     * performed during the evolution.
+     */
+    std::function<bool (const double t,
+                        VectorType &sol)> solver_should_restart;
+
+    /**
+     * Return a vector whose components are the weights used by ARKode to compute
+     * the vector norm. The implementation of this function is optional, and it
+     * is used only when `use_local_tolerances` is set to true at construction
+     * time, or through the parameter file.
+     *
+     * If you do not overwrite this function and set `use_local_tolerances` to
+     * true, an exception will be thrown when trying to start the time stepper.
+     */
+    std::function<VectorType&()> get_local_tolerances;
+
+    /**
+     * Handle ARKode exceptions.
+     */
+    DeclException1(ExcARKodeError, int, << "One of the SUNDIALS ARKode internal functions "
+                   << " returned a negative error code: "
+                   << arg1 << ". Please consult SUNDIALS manual.");
+
+
+  private:
+
+    /**
+     * Throw an exception when a function with the given name is not implemented.
+     */
+    DeclException1(ExcFunctionNotProvided, std::string,
+                   << "Please provide an implementation for the function \"" << arg1 << "\"");
+
+    /**
+     * This function is executed at construction time to set the
+     * std::function above to trigger an assert if they are not
+     * implemented.
+     */
+    void set_functions_to_trigger_an_assert();
+
+    /**
+     * ARKode configuration data.
+     */
+    AdditionalData data;
+
+    /**
+     * ARKode memory object.
+     */
+    void *arkode_mem;
+
+    /**
+     * ARKode solution vector.
+     */
+    N_Vector yy;
+
+    /**
+     * ARKode absolute tolerances vector.
+     */
+    N_Vector abs_tolls;
+
+#ifdef DEAL_II_WITH_MPI
```

Comment:
you have `MPI_Comm` in constructor, does this PR builds without MPI?

Reply:
Yes. We define `MPI_Comm` and `MPI_COMM_WORLD` as `int` and `0` in  `deal.II/base/mpi.h` if we compile without MPI.